alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java

```diff
@@ -0,0 +1,221 @@
+    }
+
+    private String getShardInstanceIdentifier(String in, int index) {
+        String[] segments = in.split("/");
+        if (index == -1) {
+            return segments[segments.length - 1];
+        } else {
+            return segments[index];
+        }
+    }
+
+    private JSONObject getTimingsFromResponseJson(JSONObject jsonResponse) {
+        if (jsonResponse != null && sharded == false) {
+            return jsonResponse.optJSONObject(DEBUG_KEY).optJSONObject(TIMINGS_KEY);
+        } else {
+            return null;
+        }
+    }
+
+    private static String findTimings(String s) {
+        Pattern timingsRegex = Pattern.compile("timing=(.*),processedDenies");
+        Matcher matcher = timingsRegex.matcher(s);
+        matcher.find();
+        return matcher.group(1);
+    }
+
+    private void createShardedRequests(JSONObject jsonResponse, JSONObject originalParams, String query) {
+        String[] shards = originalParams.getString(SHARDS_KEY).split(",");
+        shardedRequests = new HashMap<>();
+
+        //Loop over shards and add debug/tracking information to each request representation
+        for (int shardIndex = 0; shardIndex < shards.length; shardIndex++) {
+            String shardName = shards[shardIndex];
+            String shardShortName = getShardInstanceIdentifier(shardName, -1);
+
+            //Add tracking information to underlying sharded requests
+            if (jsonResponse.has(DEBUG_KEY) && jsonResponse.getJSONObject(DEBUG_KEY).has(TRACK_KEY)) {
+                JSONObject trackingDebug = jsonResponse.getJSONObject(DEBUG_KEY).getJSONObject(TRACK_KEY);
+
+                HashMap<String, ShardedSolrSubRequest> subRequestHashMap = new HashMap<>();
+                Iterator<String> keys = trackingDebug.keys();
+
+                long cummulativeQTime = 0;
+                long totalElapsed = 0;
+                Set<Integer> numFoundSet = new HashSet<>();
+
+                while (keys.hasNext()) {
+                    String key = keys.next();
+                    JSONObject value = trackingDebug.optJSONObject(key);
+                    if (!key.equals(RID_KEY) && value.has(shardName)) {
+                        //In a ShardedSolrSubRequest
+                        JSONObject subRequestTrackingInfo = value.getJSONObject(shardName);
+
+                        long srQTime = subRequestTrackingInfo.optLong(QTIME_KEY);
+                        int srNumFound = subRequestTrackingInfo.optInt(SUB_NUMFOUND_KEY);
+                        long srElapsedTime = subRequestTrackingInfo.optLong(ELAPSEDTIME_KEY);
+                        String requestPurpose = subRequestTrackingInfo.optString(REQUEST_PURPOSE_KEY);
+
+                        String subResponse = subRequestTrackingInfo.optString("Response");
+
+                        JSONObject timings;
+                        try {
+                            // Timings section in debug information is not necessarily a correctly formatted JSONObject.
```

Comment:
Do you have an example of this incorrectly formatted json?

> correctly formatted JSONObject

I believe you are talking about just the JSON string, before attempting to parse it, that is invalid JSON.
Referring to JSONObject (the Java class) is a bit confiusing here.